# Analysis of eBay Car Sales Data

We'll work with a dataset of used cars from *eBay Kleinanzeigen*, a [classified](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website. The dataset was originally scraped and uploaded to Kaggle by user [orgesleka](https://www.kaggle.com/orgesleka).
The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data). The aim of this project is to clean the data and analyze the included used car listings. 

Let's start by importing the libraries we need and reading the dataset into pandas.

In [1159]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="Windows-1252")

In [1160]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [1161]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Cleaning Column Names

Our dataset consists of 50,000 entries and 20 columns. We can see that columns such as `yearOfRegistration`, `powerPS`, `monthOfRegistration`, `nrOfPictures` and `postalCode` are represented as the int64 type. We can assume, that data in those columns is expressed by numbers. The rest of the columns are represented as the object type, indicating that they are expressed by strings, not numbers. Also, some columns (`vehicleType`, `gearbox`, `model`, `fuelType`, `notRepairedDamage`) have null values.

We can also see that the columns names use camelcase instead of Python's preferrede snakecase. Let's convert the column names from camelcase to snakecase and reword some of the column names based on the [data dictionary](https://data.world/data-society/used-cars-data) to be more descriptive.

In [1162]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [1163]:
copied_columns = autos.columns

mapping_dict = {
    'dateCrawled': 'date_crawled',
    'name': 'name',
    'seller': 'seller',
    'offerType': 'offer_type',
    'price': 'price',
    'abtest': 'abtest',
    'vehicleType': 'vehicle_type',
    'yearOfRegistration': 'registration_year',
    'gearbox': 'gearbox',
    'powerPS': 'power_ps',
    'model': 'model',
    'odometer': 'odometer',
    'monthOfRegistration': 'registration_month',
    'fuelType': 'fuel_type',
    'brand': 'brand',
    'notRepairedDamage': 'unrepaired_damage',
    'dateCreated': 'ad_created',
    'nrOfPictures': 'nr_of_pictures',
    'postalCode': 'postal_code',
    'lastSeen': 'last_seen'
}

autos.columns = copied_columns.map(mapping_dict)   # Run only once!

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We've made following changes to column names:
- `yearOfRegistration` to `registration_year`,
- `monthOfRegistration` to `registration_month`,
- `notRepairedDamage` to `unrepaired_damage`,
- `dateCreated` to `ad_created`,
- the rest of the column names from camelcase to snakecase.

Thanks to these changes column names are now in Python's preferred format and are more descriptive.

## German Words Translation

In columns `seller`, `offer_type`, `vehicle_type`, `gearbox`, `fuel_type` and `unrepaired_damage` there are german words. We can translate them and map the values to their english counterparts. 

In [1164]:
autos['seller'].unique()

array(['privat', 'gewerblich'], dtype=object)

In [1165]:
corrections_seller = {
    'gewerblich': 'commercial',
    'privat': 'private'
}

autos['seller'] = autos['seller'].map(corrections_seller)

autos['seller'].unique()

array(['private', 'commercial'], dtype=object)

In [1166]:
autos['offer_type'].unique()

array(['Angebot', 'Gesuch'], dtype=object)

In [1167]:
corrections_offer_type = {
    'Angebot': 'Request',
    'Gesuch': 'Offer'
}

autos['offer_type'] = autos['offer_type'].map(corrections_offer_type)

autos['offer_type'].unique()

array(['Request', 'Offer'], dtype=object)

In [1168]:
autos['vehicle_type'].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [1169]:
corrections_vehicle_type = {
    'bus': 'bus',
    'limousine': 'limousine',
    'kleinwagen': 'small car',
    'kombi': 'combi',
    'coupe': 'coupe',
    'suv': 'suv',
    'cabrio': 'cabriolet',
    'andere': 'other'
}

autos['vehicle_type'] = autos['vehicle_type'].map(corrections_vehicle_type)

autos['vehicle_type'].unique()

array(['bus', 'limousine', 'small car', 'combi', nan, 'coupe', 'suv',
       'cabriolet', 'other'], dtype=object)

In [1170]:
autos['gearbox'].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [1171]:
corrections_gearbox = {
    'manuell': 'manual',
    'automatik': 'automatic'
}

autos['gearbox'] = autos['gearbox'].map(corrections_gearbox)

autos['gearbox'].unique()

array(['manual', 'automatic', nan], dtype=object)

In [1172]:
autos['fuel_type'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [1173]:
corrections_fuel_type = {
    'lpg': 'lpg',
    'benzin': 'petrol',
    'diesel': 'diesel',
    'cng': 'cng',
    'hybrid': 'hybrid',
    'elektro': 'electro',
    'andere': 'other'
}

autos['fuel_type'] = autos['fuel_type'].map(corrections_fuel_type)

autos['fuel_type'].unique()

array(['lpg', 'petrol', 'diesel', nan, 'cng', 'hybrid', 'electro',
       'other'], dtype=object)

In [1174]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

In [1175]:
corrections_unrepaired_damage = {
    'nein': 'no',
    'ja': 'yes'
}

autos['unrepaired_damage'] = autos['unrepaired_damage'].map(corrections_unrepaired_damage)

autos['unrepaired_damage'].unique()

array(['no', nan, 'yes'], dtype=object)

In [1176]:
autos['model'].value_counts()

golf          4024
andere        3528
3er           2761
polo          1757
corsa         1735
              ... 
kappa            2
200              1
rangerover       1
i3               1
b_max            1
Name: model, Length: 245, dtype: int64

In [1177]:
autos.loc[autos['model'] == 'andere', 'model'] = 'other'

Now we have only english words in the dataset.

## Initial Exploration and Cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [1178]:
autos.describe(include='all')

date_crawled         name   seller offer_type  price abtest  \
count                 50000        50000    50000      50000  50000  50000   
unique                48213        38754        2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  private    Request     $0   test   
freq                      3           78    49999      49999   1421  25756   
mean                    NaN          NaN      NaN        NaN    NaN    NaN   
std                     NaN          NaN      NaN        NaN    NaN    NaN   
min                     NaN          NaN      NaN        NaN    NaN    NaN   
25%                     NaN          NaN      NaN        NaN    NaN    NaN   
50%                     NaN          NaN      NaN        NaN    NaN    NaN   
75%                     NaN          NaN      NaN        NaN    NaN    NaN   
max                     NaN          NaN      NaN        NaN    NaN    NaN   

       vehicle_type  registration_year gearbox      power_ps  model  \
count         44905       50000.000000   47320  50000.000000  47242   
unique            8                NaN       2           NaN    245   
top       limousine                NaN  manual           NaN   golf   
freq          12859                NaN   36993           NaN   4024   
mean            NaN        2005.073280     NaN    116.355920    NaN   
std             NaN         105.712813     NaN    209.216627    NaN   
min             NaN        1000.000000     NaN      0.000000    NaN   
25%             NaN        1999.000000     NaN     70.000000    NaN   
50%             NaN        2003.000000     NaN    105.000000    NaN   
75%             NaN        2008.000000     NaN    150.000000    NaN   
max             NaN        9999.000000     NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    petrol  volkswagen                no   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

As we can see, all main statistics for `nr_of_pictures` column are equal to 0. We can assume that this column have only one value, therefore it's irrelevant for the purposes of the analysis.

In [1179]:
autos.drop('nr_of_pictures', inplace=True, axis=1)

autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

        seller offer_type    price   abtest vehicle_type  registration_year  \
0      private    Request   $5,000  control          bus               2004   
1      private    Request   $8,500  control    limousine               1997   
2      private    Request   $8,990     test    limousine               2009   
3      private    Request   $4,350  control    small car               2007   
4      private    Request   $1,350     test        combi               2003   
...        ...        ...      ...      ...          ...                ...   
49995  private    Request  $24,900  control    limousine               2011   
49996  private    Request   $1,980  control    cabriolet               1996   
49997  private    Request  $13,200     test    cabriolet               2014   
49998  private    Request  $22,900  control        combi               2013   
49999  private    Request   $1,250  control    limousine               1996   

         gearbox  power_ps   model   odometer  registration_month fuel_type  \
0         manual       158   other  150,000km                   3       lpg   
1      automatic       286     7er  150,000km                   6    petrol   
2         manual       102    golf   70,000km                   7    petrol   
3      automatic        71  fortwo   70,000km                   6    petrol   
4         manual         0   focus  150,000km                   7    petrol   
...          ...       ...     ...        ...                 ...       ...   
49995  automatic       239      q5  100,000km                   1    diesel   
49996     manual        75   astra  150,000km                   5    petrol   
49997  automatic        69     500    5,000km                  11    petrol   
49998     manual       150      a3   40,000km                  11    diesel   
49999     manual       101  vectra  150,000km                   1    petrol   

            brand unrepaired_damage           ad_created  postal_code  \
0         peugeot                no  2016-03-26 00:00:00        79588   
1             bmw                no  2016-04-04 00:00:00        71034   
2      volkswagen                no  2016-03-26 00:00:00        35394   
3           smart                no  2016-03-12 00:00:00        33729   
4            ford                no  2016-04-01 00:00:00        39218   
...           ...               ...                  ...          ...   
49995        audi                no  2016-03-27 00:00:00        82131   
49996        opel                no  2016-03-28 00:00:00        44807   
49997        fiat                no  2016-04-02 00:00:00        73430   
49998        audi                no  2016-03-08 00:00:00        35683   
49999        opel                no  2016-03-13 00:00:00        45897   

                 last_seen  
0      2016-04-06 06:45:54  
1      2016-04-06 14:45:08  
2      2016-04-06 20:15:37  
3      2016-03-15 03:16:28  
4      2016-04-01 14:38:50  
...                    ...  
49995  2016-04-01 13:47:40  
49996  2016-04-02 14:18:02  
49997

Columns such as `seller`, `offer_type`, `abtest`, `gearbox` and `unrepaired_damage` have two unique values. These columns need further investigation. If these columns have mostly one value, they're candidates to be dropped.

Data in `price` column is numeric data stored as text data because of the `$` and `,` characters. The same is true for `odometer` column, but this time because of `,` character and `km` string.

In [1180]:
print(autos['seller'].value_counts())
print('\n')
print(autos['offer_type'].value_counts())
print('\n')
print(autos['abtest'].value_counts())
print('\n')
print(autos['gearbox'].value_counts())
print('\n')
print(autos['unrepaired_damage'].value_counts())

private       49999
commercial        1
Name: seller, dtype: int64


Request    49999
Offer          1
Name: offer_type, dtype: int64


test       25756
control    24244
Name: abtest, dtype: int64


manual       36993
automatic    10327
Name: gearbox, dtype: int64


no     35232
yes     4939
Name: unrepaired_damage, dtype: int64


After further investigation we can see, that `seller` column have mostly one value (only 1 `commercial` value and 49,999 `private` values). Furthermore, `offer_type` column also have mostly one value (only 1 `Offer` value and 49,999 `Request` values). Those two columns are additional candidates to be dropped.

Two entries containing these unique values for both columns are shown below.

In [1181]:
autos[(autos['seller'] == 'commercial') | (autos['offer_type'] == 'Offer')]

date_crawled                                         name  \
7738   2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   
17541  2016-04-03 15:48:33                         Suche_VW_T5_Multivan   

           seller offer_type price   abtest vehicle_type  registration_year  \
7738   commercial    Request  $100  control        combi               2000   
17541     private      Offer    $0     test          bus               2005   

      gearbox  power_ps        model   odometer  registration_month fuel_type  \
7738   manual         0       megane  150,000km                   8    petrol   
17541     NaN         0  transporter  150,000km                   0       NaN   

            brand unrepaired_damage           ad_created  postal_code  \
7738      renault               NaN  2016-03-15 00:00:00        65232   
17541  volkswagen               NaN  2016-04-03 00:00:00        29690   

                 last_seen  
7738   2016-04-06 17:15:37  
17541  2016-04-05 15:16:06

As we mentioned, `price` and `odometer` columns are numeric values stored as text. Let's convert these columns to numeric.

In [1182]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)           # Run only once!

autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '').astype(int)    # Run only once!
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)                                 # Run only once!

autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 19 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int64 
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer_km         50000 non-null  int64 
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

## Exploring the Odometer and Price Columns

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the `odometer_km` and `price` columns.

In [1183]:
autos['price'].unique().shape

(2357,)

In [1184]:
autos['odometer_km'].unique().shape

(13,)

There are 2357 unique values in `price` column and 13 unique values in `odometer_km` column.

In [1185]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [1186]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

We can see that in `price` column minimum value is equal to `0`, and maximum value is equal to `100,000,000`. Those values are outliers in our dataset. It's impossible for any car to cost `$0`. Moreover, `$100,000,000` is definitely too high for a car on eBay.

In [1187]:
autos['price'].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

In [1188]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [1189]:
autos['odometer_km'].value_counts().sort_index(ascending=False).head(20)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

After further investigation we can see that in `price` column value `0` occurs 1,421 times. This is a significant quantity, so we can assume, that value `0` isn't an outlier in out dataset. On the other hand, there are some extremely high values. Our decision is to remove values greater than 350,000. Such prices are exceptionally high for a car on eBay auction.

No outliers are observed in the `odometer_km` column.

In [1190]:
autos = autos[autos['price'].between(0, 350000)]

In [1191]:
autos['price'].describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price, dtype: float64

We've removed 14 entries from our dataset. Now maximum value in `price` column is 350,000. This operation significantly influenced the mean of `price` column. The mean decreased from 9,840 to about 5,722. This is huge difference. The rest of the statistics (first quartile, mean, third quartile) haven't changed at all. This is proof that the decision to remove outliers was accurate.

## Exploring the Date Columns

Let's now move on to the date columns and understand the date range the data covers. There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [1192]:
autos[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The first 10 characters represent the day (e.g. `2016-03-26`). To understand the date range, we can extract just the date values, use `Series.value_counts()` to generate a distribution, and then sort by the index.

To select the first 10 characters in each column, we can use `Series.str[:10]`:

In [1193]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 49986, dtype: object


Firstly, let's calculate the distribution of values in the `date_crawled` column.

In [1194]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False)

2016-04-03    0.038691
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-12    0.036770
2016-03-14    0.036630
2016-04-04    0.036490
2016-03-07    0.035970
2016-04-02    0.035410
2016-03-19    0.034910
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-15    0.033990
2016-04-01    0.033809
2016-03-30    0.033629
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-22    0.032909
2016-03-11    0.032489
2016-03-26    0.032489
2016-03-23    0.032389
2016-03-10    0.032129
2016-03-31    0.031909
2016-03-25    0.031749
2016-03-17    0.031509
2016-03-27    0.031049
2016-03-16    0.029508
2016-03-24    0.029108
2016-03-05    0.025387
2016-03-13    0.015564
2016-03-06    0.013944
2016-04-05    0.013104
2016-03-18    0.013064
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

In [1195]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
2016-03-25    0.031749
2016-03-26    0.032489
2016-03-27    0.031049
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033629
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035410
2016-04-03    0.038691
2016-04-04    0.036490
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

The site consistently crawled data each day from `2016-03-05` to `2016-04-07`. The amount of crawled data is irregular: the most data was crawled on `2016-04-03` (3.87%), and on the other hand the least data was crawled on `2016-04-07` (0.14%).

Now let's calculate the distribution of values in the `ad_created` column.

In [1196]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False)

2016-04-03    0.038931
2016-03-20    0.037871
2016-03-21    0.037691
2016-04-04    0.036850
2016-03-12    0.036610
                ...   
2016-01-29    0.000020
2016-02-17    0.000020
2016-01-14    0.000020
2015-09-09    0.000020
2016-02-11    0.000020
Name: ad_created, Length: 76, dtype: float64

In [1197]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index().head(40)

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
2015-12-30    0.00002
2016-01-03    0.00002
2016-01-07    0.00002
2016-01-10    0.00004
2016-01-13    0.00002
2016-01-14    0.00002
2016-01-16    0.00002
2016-01-22    0.00002
2016-01-27    0.00006
2016-01-29    0.00002
2016-02-01    0.00002
2016-02-02    0.00004
2016-02-05    0.00004
2016-02-07    0.00002
2016-02-08    0.00002
2016-02-09    0.00004
2016-02-11    0.00002
2016-02-12    0.00006
2016-02-14    0.00004
2016-02-16    0.00002
2016-02-17    0.00002
2016-02-18    0.00004
2016-02-19    0.00006
2016-02-20    0.00004
2016-02-21    0.00006
2016-02-22    0.00002
2016-02-23    0.00008
2016-02-24    0.00004
2016-02-25    0.00006
2016-02-26    0.00004
2016-02-27    0.00012
2016-02-28    0.00020
2016-02-29    0.00016
2016-03-01    0.00010
2016-03-02    0.00010
Name: ad_created, dtype: float64

In [1198]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index().tail(40)

2016-02-28    0.000200
2016-02-29    0.000160
2016-03-01    0.000100
2016-03-02    0.000100
2016-03-03    0.000860
2016-03-04    0.001440
2016-03-05    0.023046
2016-03-06    0.015124
2016-03-07    0.034750
2016-03-08    0.033309
2016-03-09    0.033229
2016-03-10    0.031869
2016-03-11    0.032789
2016-03-12    0.036610
2016-03-13    0.016925
2016-03-14    0.035230
2016-03-15    0.033749
2016-03-16    0.030008
2016-03-17    0.031189
2016-03-18    0.013724
2016-03-19    0.033849
2016-03-20    0.037871
2016-03-21    0.037691
2016-03-22    0.032769
2016-03-23    0.032189
2016-03-24    0.029088
2016-03-25    0.031889
2016-03-26    0.032569
2016-03-27    0.030909
2016-03-28    0.034970
2016-03-29    0.034110
2016-03-30    0.033449
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035090
2016-04-03    0.038931
2016-04-04    0.036850
2016-04-05    0.011843
2016-04-06    0.003261
2016-04-07    0.001280
Name: ad_created, dtype: float64

In our dataset insignificant number of ads on eBay were created before `2016-03-05`. Starting of `2016-03-05` (the day the data started to be crawled), up to `2016-04-04`, similar amounts of ads were created each day. This period almost perfectly overlaps with the time when the data was crawled.

In [1199]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False)

2016-04-06    0.220982
2016-04-07    0.130957
2016-04-05    0.124275
2016-03-17    0.027928
2016-04-03    0.025367
2016-04-02    0.024887
2016-03-30    0.024847
2016-04-04    0.024627
2016-03-31    0.023827
2016-03-12    0.023807
2016-04-01    0.023106
2016-03-29    0.022326
2016-03-22    0.021586
2016-03-28    0.020846
2016-03-21    0.020726
2016-03-20    0.020706
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-23    0.018585
2016-03-26    0.016965
2016-03-16    0.016445
2016-03-27    0.016024
2016-03-15    0.015884
2016-03-19    0.015744
2016-03-14    0.012804
2016-03-11    0.012524
2016-03-10    0.010763
2016-03-09    0.009843
2016-03-13    0.008983
2016-03-08    0.007582
2016-03-18    0.007422
2016-03-07    0.005362
2016-03-06    0.004421
2016-03-05    0.001080
Name: last_seen, dtype: float64

In [1200]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-26    0.016965
2016-03-27    0.016024
2016-03-28    0.020846
2016-03-29    0.022326
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023106
2016-04-02    0.024887
2016-04-03    0.025367
2016-04-04    0.024627
2016-04-05    0.124275
2016-04-06    0.220982
2016-04-07    0.130957
Name: last_seen, dtype: float64

We can see that date range in `last_seen` column is the same as the date range in `date_crawled` column. Most ads were last seen by the crawler in the last three days:
- `2016-04-05`: 12.43%
- `2016-04-06`: 22.10%
- `2016-04-07`: 13.10%

Now let's move to the distribution of `registration_year` column:

In [1201]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [1202]:
autos['registration_year'].value_counts(dropna=False).sort_index().head(10)

1000    1
1001    1
1111    1
1500    1
1800    2
1910    9
1927    1
1929    1
1931    1
1934    2
Name: registration_year, dtype: int64

In [1203]:
autos['registration_year'].value_counts(dropna=False).sort_index().tail(20)

2011    1634
2012    1323
2013     806
2014     665
2015     399
2016    1316
2017    1452
2018     491
2019       3
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       2
9996       1
9999       4
Name: registration_year, dtype: int64

There are some incorrect values in `registration_year` column of our dataset. It's impossible for any car to be registered in 1800 or earlier (many years before cars were invented). There are also some values many years into the future (e.g. 2800, 4100, 9999, etc.).

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above `2016` is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [1204]:
autos = autos[autos['registration_year'].between(1900, 2016)]

autos['registration_year'].value_counts(normalize=True)

2000    0.069852
2005    0.062792
1999    0.062438
2004    0.057002
2003    0.056794
          ...   
1939    0.000021
1927    0.000021
1929    0.000021
1948    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

In [1205]:
autos['registration_year'].value_counts(normalize=True).head(10)

2000    0.069852
2005    0.062792
1999    0.062438
2004    0.057002
2003    0.056794
2006    0.056377
2001    0.056273
2002    0.052753
1998    0.051087
2007    0.047984
Name: registration_year, dtype: float64

In [1206]:
autos['registration_year'].value_counts(normalize=True).tail(10)

1955    0.000042
1931    0.000021
1953    0.000021
1943    0.000021
1938    0.000021
1939    0.000021
1927    0.000021
1929    0.000021
1948    0.000021
1952    0.000021
Name: registration_year, dtype: float64

Most cars were registered in `2000` (almost 7%). In general we can observe that most cars were registered around the turn of the twentieth and twenty-first centuries. Fewest cars were registered long time ago, in the first half of the twentieth century.

## Exploring Price by Brand

Now we'll use aggregation to understand the `brand` column.

In [1207]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212117
bmw               0.110026
opel              0.108172
mercedes_benz     0.095364
audi              0.086409
ford              0.069768
renault           0.047359
peugeot           0.029532
fiat              0.025866
seat              0.018181
skoda             0.016036
mazda             0.015141
nissan            0.015099
smart             0.013912
citroen           0.013912
toyota            0.012475
sonstige_autos    0.010892
hyundai           0.009851
volvo             0.009247
mini              0.008643
mitsubishi        0.008143
honda             0.007852
kia               0.007102
alfa_romeo        0.006623
porsche           0.006102
suzuki            0.005915
chevrolet         0.005706
chrysler          0.003665
dacia             0.002562
daihatsu          0.002562
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001562
daewoo            0.001500
r

In [1208]:
top_brands = autos['brand'].value_counts(normalize=True) >= 0.02
top_brands[top_brands].index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat'],
      dtype='object')

In `brand` column there are:
- `volkswagen` cars: 21.2%,
- `bmw` cars: 11.0%,
- `opel` cars: 10.8%,
- `mercedes_benz` cars: 9.5%,
- `audi` cars: 8.6%,
- `ford` cars: 7.0%,
- `renault` cars: 4.7%,
- `peugeot` cars: 3.0%,
- `fiat` cars: 2.6%.

All other car brands account for less than 2% of the share in the dataset. We've choosen all brands that have over 2% of the total values to aggregate on.

Now we'll create an empty dictionary to hold our aggregate data. We'll loop over selected brands, and assign the mean price to the dictionary, with the brand name as the key.

In [1209]:
brands_index = top_brands[top_brands].index

brand_mean_prices  = {}

for brand in brands_index:
    price_mean = autos.loc[autos['brand'] == brand, 'price'].mean()
    brand_mean_prices [brand] = price_mean

print(brand_mean_prices)

{'volkswagen': 5231.081983308787, 'bmw': 8102.536248343744, 'opel': 2876.716403542549, 'mercedes_benz': 8485.239571958942, 'audi': 9093.65003615329, 'ford': 3652.095223880597, 'renault': 2395.4164467897976, 'peugeot': 3039.4682651622, 'fiat': 2711.8011272141707}


In [1210]:
import operator

sorted_by_price = sorted(brand_mean_prices .items(), key=operator.itemgetter(1), reverse=True)

brand_mean_prices  = dict(sorted_by_price)

print(brand_mean_prices )

{'audi': 9093.65003615329, 'mercedes_benz': 8485.239571958942, 'bmw': 8102.536248343744, 'volkswagen': 5231.081983308787, 'ford': 3652.095223880597, 'peugeot': 3039.4682651622, 'opel': 2876.716403542549, 'fiat': 2711.8011272141707, 'renault': 2395.4164467897976}


Our analysis shows that (on average) among the top 9 brands, the most expensive are:
- Audi: \$9,094
- Mercedes Benz: \$8,485
- BMW: \$8,103

Average prices of those cars are above mean car price in whole dataset (\$5,722).

On average, less expensive are:
- Ford: \$3,652
- Peugeot: \$3,039
- Opel: \$2,877
- Renault: \$2,395.

Volkswagen, with average price $5,231, is somewhere in between.

## Mileage Analysis for the Top 9 Brands

Now let's use aggregation to understand the average mileage for cars from top 9 brands and if there's any visible link with mean price. We can combine the data from mean price and mileage series objects into a single dataframe (with a shared index) and display the dataframe directly. Again we'll use the loop method to calculate the mean mileage for each of the top brands and store the results in a dictionary.

In [1211]:
brand_mean_mileage  = {}

for brand in brands_index:
    mileage_mean = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    brand_mean_mileage [brand] = mileage_mean

print(brand_mean_mileage)

{'volkswagen': 128724.10407461954, 'bmw': 132431.38368351315, 'opel': 129223.14208702349, 'mercedes_benz': 130856.0821139987, 'audi': 129287.78018799711, 'ford': 124068.65671641791, 'renault': 128183.81706244503, 'peugeot': 127136.81241184767, 'fiat': 116553.94524959743}


Now we'll convert price and mileage dictionaries to series objects, using the series constructor, and then we'll create a dataframe from the first series object using the dataframe constructor.

In [1212]:
price_series = pd.Series(brand_mean_prices)
mileage_series = pd.Series(brand_mean_mileage)

price_mileage_df = pd.DataFrame(price_series, columns=['mean_price'])
price_mileage_df['mean_mileage'] = mileage_series

price_mileage_df.sort_values('mean_price', ascending=False)

mean_price   mean_mileage
audi           9093.650036  129287.780188
mercedes_benz  8485.239572  130856.082114
bmw            8102.536248  132431.383684
volkswagen     5231.081983  128724.104075
ford           3652.095224  124068.656716
peugeot        3039.468265  127136.812412
opel           2876.716404  129223.142087
fiat           2711.801127  116553.945250
renault        2395.416447  128183.817062

The average mileage values for each brand are in the range 116,000 - 133,000. As we can see, there isn't any clear pattern in the mean price values, given the mean mileage values.

## Dates Conversion from String to Integers

Let's convert the dates to be uniform numeric data, so `"2016-03-21"` becomes the integer `20160321`.

In [1213]:
autos['date_crawled_time'] = autos['date_crawled'].str.split().str[1]
autos['date_crawled'] = autos['date_crawled'].str.split().str[0].str.replace('-', '').astype(int)

autos['ad_created_time'] = autos['ad_created'].str.split().str[1]
autos['ad_created'] = autos['ad_created'].str.split().str[0].str.replace('-', '').astype(int)

autos['last_seen_time'] = autos['last_seen'].str.split().str[1]
autos['last_seen'] = autos['last_seen'].str.split().str[0].str.replace('-', '').astype(int)

autos = autos[[
    'date_crawled',
    'date_crawled_time',
    'name',
    'seller',
    'offer_type',
    'price',
    'abtest',
    'vehicle_type',
    'registration_year',
    'gearbox',
    'power_ps',
    'model',
    'odometer_km',
    'registration_month',
    'fuel_type',
    'brand',
    'unrepaired_damage',
    'ad_created',
    'ad_created_time',
    'postal_code',
    'last_seen',
    'last_seen_time'
    ]]

autos

date_crawled date_crawled_time  \
0          20160326          17:47:46   
1          20160404          13:38:56   
2          20160326          18:57:24   
3          20160312          16:58:10   
4          20160401          14:38:50   
...             ...               ...   
49995      20160327          14:38:19   
49996      20160328          10:50:25   
49997      20160402          14:44:48   
49998      20160308          19:25:42   
49999      20160314          00:42:12   

                                                    name   seller offer_type  \
0                       Peugeot_807_160_NAVTECH_ON_BOARD  private    Request   
1             BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private    Request   
2                             Volkswagen_Golf_1.6_United  private    Request   
3      Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private    Request   
4      Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private    Request   
...                                                  ...      ...        ...   
49995   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon  private    Request   
49996  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...  private    Request   
49997                     Fiat_500_C_1.2_Dualogic_Lounge  private    Request   
49998                 Audi_A3_2.0_TDI_Sportback_Ambition  private    Request   
49999                                Opel_Vectra_1.6_16V  private    Request   

       price   abtest vehicle_type  registration_year    gearbox  ...  \
0       5000  control          bus               2004     manual  ...   
1       8500  control    limousine               1997  automatic  ...   
2       8990     test    limousine               2009     manual  ...   
3       4350  control    small car               2007  automatic  ...   
4       1350     test        combi               2003     manual  ...   
...      ...      ...          ...                ...        ...  ...   
49995  24900  control    limousine               2011  automatic  ...   
49996   1980  control    cabriolet               1996     manual  ...   
49997  13200     test    cabriolet               2014  automatic  ...   
49998  22900  control        combi               2013     manual  ...   
49999   1250  control    limousine               1996     manual  ...   

       odometer_km registration_month  fuel_type       brand  \
0           150000                  3        lpg     peugeot   
1           150000                  6     petrol         bmw   
2            70000                  7     petrol  volkswagen   
3            70000                  6     petrol       smart   
4           150000                  7     petrol        ford   
...            ...                ...        ...         ...   
49995       100000                  1     diesel        audi   
49996       150000                  5     petrol        opel   
49997         5000                 11     petrol        fiat   
49998        40000                 11     diesel        audi   
49999       150000                  1     petrol        opel   

      unrepaired_damage ad_created ad_created_time  postal_code last_seen  \
0                    no   20160326        00:00:00        79588  20160406   
1                    no   20160404        00:00:00        71034  20160406   
2                    no   20160326        00:00:00        35394  20160406   
3                    no   20160312        00:00:00        33729  20160315   
4                    no   20160401        00:00:00        39218  20160401   
...                 ...        ...             ...          ...       ...   
49995                no   20160327        00:00:00        82131  20160401   
49996                no   20160328        00:00:00        44807  20160402   
49997                no   20160402        00:00:00        73430  20160404   
49998                no   20160308        00:00:00        35683  20160405   
49999                no   20160313        00:00:00        45

In [1214]:
autos['ad_created_time'].unique()

array(['00:00:00'], dtype=object)

All values in `ad_created_time` column are `00:00:00`, so we can drop this column.

In [1215]:
autos.drop('ad_created_time', inplace=True, axis=1)

autos

date_crawled date_crawled_time  \
0          20160326          17:47:46   
1          20160404          13:38:56   
2          20160326          18:57:24   
3          20160312          16:58:10   
4          20160401          14:38:50   
...             ...               ...   
49995      20160327          14:38:19   
49996      20160328          10:50:25   
49997      20160402          14:44:48   
49998      20160308          19:25:42   
49999      20160314          00:42:12   

                                                    name   seller offer_type  \
0                       Peugeot_807_160_NAVTECH_ON_BOARD  private    Request   
1             BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private    Request   
2                             Volkswagen_Golf_1.6_United  private    Request   
3      Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private    Request   
4      Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private    Request   
...                                                  ...      ...        ...   
49995   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon  private    Request   
49996  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...  private    Request   
49997                     Fiat_500_C_1.2_Dualogic_Lounge  private    Request   
49998                 Audi_A3_2.0_TDI_Sportback_Ambition  private    Request   
49999                                Opel_Vectra_1.6_16V  private    Request   

       price   abtest vehicle_type  registration_year    gearbox  ...   model  \
0       5000  control          bus               2004     manual  ...   other   
1       8500  control    limousine               1997  automatic  ...     7er   
2       8990     test    limousine               2009     manual  ...    golf   
3       4350  control    small car               2007  automatic  ...  fortwo   
4       1350     test        combi               2003     manual  ...   focus   
...      ...      ...          ...                ...        ...  ...     ...   
49995  24900  control    limousine               2011  automatic  ...      q5   
49996   1980  control    cabriolet               1996     manual  ...   astra   
49997  13200     test    cabriolet               2014  automatic  ...     500   
49998  22900  control        combi               2013     manual  ...      a3   
49999   1250  control    limousine               1996     manual  ...  vectra   

      odometer_km  registration_month  fuel_type       brand  \
0          150000                   3        lpg     peugeot   
1          150000                   6     petrol         bmw   
2           70000                   7     petrol  volkswagen   
3           70000                   6     petrol       smart   
4          150000                   7     petrol        ford   
...           ...                 ...        ...         ...   
49995      100000                   1     diesel        audi   
49996      150000                   5     petrol        opel   
49997        5000                  11     petrol        fiat   
49998       40000                  11     diesel        audi   
49999      150000                   1     petrol        opel   

      unrepaired_damage ad_created  postal_code  last_seen  last_seen_time  
0                    no   20160326        79588   20160406        06:45:54  
1                    no   20160404        71034   20160406        14:45:08  
2                    no   20160326        35394   20160406        20:15:37  
3                    no   20160312        33729   20160315        03:16:28  
4                    no   20160401        39218   20160401        14:38:50  
...                 ...        ...          ...        ...             ...  
49995                no   20160327        82131   20160401        13:47:40  
49996                no   20160328        44807   20160402        14:18:02  
49997                no   20160402        73430   20160404        11:47:27  
49998                no   20160308       

## Most Common Brand/Model Combinations

In [1216]:
autos.fillna('no_information', inplace=True)

In [1217]:
brand_model = autos['brand'] + ' ' + autos['model']
brand_model

0          peugeot other
1                bmw 7er
2        volkswagen golf
3           smart fortwo
4             ford focus
              ...       
49995            audi q5
49996         opel astra
49997           fiat 500
49998            audi a3
49999        opel vectra
Length: 48016, dtype: object

In [1218]:
brand_model.value_counts().head(10)

volkswagen golf           3815
bmw 3er                   2688
volkswagen polo           1677
opel corsa                1645
opel astra                1388
volkswagen passat         1388
audi a4                   1265
bmw 5er                   1163
mercedes_benz c_klasse    1147
mercedes_benz e_klasse     981
dtype: int64

In [1219]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [1231]:
brand_list = autos['brand'].unique()
model_series = []
count_series = []


for brand in brand_list:
    model_count = autos.loc[(autos['brand'] == brand) & (autos['model'] != 'other'), 'model'].value_counts()
    count = model_count[0]
    count_series.append(count)
    
    model = model_count.index[0]
    model_series.append(model)

common_brand_model = pd.DataFrame(brand_list, columns=['brand'])
common_brand_model['model'] = model_series
common_brand_model['count'] = count_series

common_brand_model.sort_values('count', ascending=False)

brand           model  count
2       volkswagen            golf   3815
1              bmw             3er   2688
11            opel           corsa   1645
9             audi              a4   1265
8    mercedes_benz        c_klasse   1147
4             ford           focus    776
7          renault          twingo    636
0          peugeot         2_reihe    611
3            smart          fortwo    552
10  sonstige_autos  no_information    523
23            fiat           punto    431
6             seat           ibiza    335
24           skoda         octavia    305
14            mini          cooper    274
17          nissan           micra    245
29         hyundai         i_reihe    176
12           mazda         3_reihe    172
30           honda           civic    171
13         porsche             911    143
15          toyota           yaris    127
21      mitsubishi            colt    113
27         citroen        berlingo    102
20           volvo             v70     95
35      alfa_romeo             156     92
32          suzuki           swift     84
38        daihatsu           cuore     66
33         trabant             601     54
26             kia         sorento     53
5         chrysler         voyager     50
16           dacia         sandero     44
18            jeep           grand     43
28       chevrolet           matiz     34
25          subaru          legacy     31
34      land_rover      freelander     31
31          daewoo           matiz     23
22          jaguar          x_type     21
39          lancia         ypsilon     21
19            saab             900     16
36            lada            niva     15
37           rover  no_information      4